vmm/task/src/youki.rs

```diff
@@ -0,0 +1,646 @@
+            opts.merge_from(&mut input)?;
+        }
+        if opts.compute_size() > 0 {
+            debug!("create options: {:?}", &opts);
+        }
+
+        let id = req.id();
+
+        let stdio = match read_io(&bundle, req.id(), None).await {
+            Ok(io) => Stdio::new(&io.stdin, &io.stdout, &io.stderr, io.terminal),
+            Err(_) => Stdio::new(req.stdin(), req.stdout(), req.stderr(), req.terminal()),
+        };
+
+        // for qemu, the io path is pci address for virtio-serial
+        // that needs to be converted to the serial file path
+        let stdio = convert_stdio(&stdio).await?;
+
+        let init = self.do_create(id, &stdio, &opts, &bundle).await?;
+        let container = YoukiContainer {
+            id: id.to_string(),
+            bundle: bundle.to_string(),
+            init,
+            process_factory: YoukiExecFactory {
+                bundle: bundle.to_string(),
+                io_uid: opts.io_uid,
+                io_gid: opts.io_gid,
+            },
+            processes: Default::default(),
+        };
+        Ok(container)
+    }
+
+    async fn cleanup(&self, _ns: &str, c: &YoukiContainer) -> containerd_shim::Result<()> {
+        self.sandbox.lock().await.defer_storages(&c.id).await?;
+        Ok(())
+    }
+}
+
+impl YoukiFactory {
+    pub fn new(sandbox: Arc<Mutex<SandboxResources>>) -> Self {
+        Self { sandbox }
+    }
+
+    async fn do_create(
+        &self,
+        id: &str,
+        stdio: &Stdio,
+        opts: &Options,
+        bundle: &str,
+    ) -> Result<InitProcess> {
+        // youki seems not support no_pivot_root and no_new_keyring option yet.
+
+        let (socket, pio, _container_io) = if stdio.terminal {
+            let s = ConsoleSocket::new().await?;
+            (Some(s), None, None)
+        } else {
+            let (pio, container_io) = create_io(id, opts.io_uid, opts.io_gid, stdio)?;
+            (None, Some(pio), Some(container_io))
+        };
+        let id_clone = id.to_string();
+        let opts_clone = opts.clone();
+        let bundle_clone = bundle.to_string();
+        let socket_path = socket.as_ref().map(|p| p.path.clone());
+        let resp = spawn_blocking(move || {
+            let builder = ContainerBuilder::new(id_clone, SyscallType::default())
+                .with_console_socket(socket_path)
+                .with_root_path(PathBuf::from(YOUKI_DIR))
+                .map_err(other_error!(e, "failed to set youki create root path"))?;
+            // if let Some(f) = container_io {
+            //     if let Some(p) = f.stdin {
+            //         builder = builder.with_stdin(p);
+            //     }
+            //     if let Some(p) = f.stdout {
+            //         builder = builder.with_stdout(p);
+            //     }
+            //     if let Some(p) = f.stderr {
+            //         builder = builder.with_stderr(p);
+            //     }
+            // }
```

Comment:
Debug code? Otherwise you need to left some comments

Reply:
done